Hex20 Nila 3U Satellite Decoder

PrajinIr opened the pull request:
A decoder for Nila, an amateur 3U satellite developed by Hex20, has been developed and tested. Its associated .yml file has been included with the directory. Corresponding changes are also implemented in CMakeLists and  __init__.py files.

daniestevez commented on python/satyaml/NILA.yml:
```diff
@@ -0,0 +1,14 @@
+    telemetry: nila
+transmitters:
+  9k6 FSK downlink:
+    frequency: 435.400e+6
```
I didn't see an entry for this transmitter [in SatNOGS DB](https://db.satnogs.org/satellite/LNSX-3317-6241-3552-3606#transmitters) and I didn't see this satellite in the [IARU satellite coordination results](https://iaru.amsat-uk.org/finished.php).

PrajinIr replied:
We are in the process of adding the transmitter in the SatNOGS community. 

Regarding IARU coordination, at the time of our application we were not fully aware of the utilization of the frequency spectrum that we selected. We are trying to resolve this for our upcoming missions.

daniestevez commented on python/telemetry/CMakeLists.txt:
```diff
@@ -80,6 +80,7 @@ GR_PYTHON_INSTALL(
     trisat.py
     upmsat_2.py
     vzlusat_2.py
+    nila.py
```
The entries here should be in alphabetic order.

PrajinIr replied:
Changed as per the order

daniestevez commented on python/telemetry/__init__.py:
```diff
@@ -63,5 +63,6 @@
 from .trisat import trisat
 from .upmsat_2 import upmsat_2
 from .vzlusat_2 import vzlusat_2
+from .nila import nila
```
The entries here should be in alphabetic order.

PrajinIr replied:
We have changed based on the order.

daniestevez replied:
It looks like a push to the PR branch with these updates is missing at the moment.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+##################################################################
+#     This Decoder File Belongs to HEX20 Labs India Pvt Ltd      #
+#     Decoder for NILA 3U Satellite                              #
+##################################################################
```
Can you follow the same header format as in other files? Also, a copyright statement and SPDX-License-Identifier should be included in the header. Code submitted to gr-satellites should be compatible with GPL-3.0-or-later.

PrajinIr replied:
We have gone through the other decoders, and we shall make the change in our decoder accordingly.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+    If, Int16sl, Int16ul, Int32sl, Int32ul, Int8sl, \
+    Int8ub, Int8ul, Padding, Struct, Switch
+
+pannel_volt_poly = [0.000000, 0.001]
+pannel_current_poly = [0.000000, 0.0005]
+
+subsystem_current_poly = [0.000000, 0.0005]
+subsystem_volt_poly = [0.000000, 0.001]
+
+aux_current_poly = [0.000000, 0.0005]
+aux_volt_poly = [0.000000, 0.001]
```
You do not need a polynomial adapter for these. The `LinearAdapter` should be used instead.

PrajinIr replied:
We have used LinearAdapter as suggested and tested.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+subsystem_temp_poly = [91.394, -0.08949, 3.55e-05, -6.26e-09, 1.89e-13]
+
+nila_header = Struct(
+    'sync_word_1' / Int8ul,
+    'sync_word_2' / Int8ul,
+    'sync_word_3' / Int8ul,
```
Does the packet include syncwords inside an AX.25 frame? That is mostly strange.

PrajinIr replied:
Thank you for pointing this out.
Yes, we have included a sync word before the CCSDS packet structure. It was a mistake from our part due to less understanding of the frame structure. This shall be taken care of in the upcoming missions.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+    'version' / Int8ul,
+    'ap_id' / Int8ul,
+    'seq_no' / Int16ul,
+    'pkt_len' / Int16ul,
```
Why is there a packet length field? AX.25 frames are delimited by HDLC, so an explicit packet length field is not needed.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+    'sat_curr_mode' / Enum(Int8ul, SAFE=0, NOMINAL=1),
+    'safe_exit_thresh' / Int16ul,
+    'safe_enter_thresh' / Int16ul,
+    'battchg_volt' / PolynomialAdapter([0.000000, 0.00125], Int16ul),
+    'battchg_curr' / PolynomialAdapter([0.000000, 0.00125], Int16ul),
```
`LinearAdapter` should be used for these.

PrajinIr replied:
We have used LinaerAdapter as suggested.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,237 @@
+            (0x60): nila_pl_pck,
+        }
+    ),
+    'crc' / Int16ul,
```
Is there a CRC-16 in addition to the CRC-16 provided by AX.25?

PrajinIr replied:
Yes, we have an additional CRC packet to ensure data integrity.

daniestevez commented on test.sh:
```diff
@@ -2,6 +2,9 @@
 
 export GR_SATELLITES_SUBMIT_TLM=0
 
+
+echo Nila
+gr_satellites nila --wavfile satellite-recordings/Nila.wav
```
To be tested when the corresponding recording is submitted to satellite-recordings.

PrajinIr replied:
After cloning the repository, we did not find any recording in the recordings folder. Our recording was added in local but couldn't be merged. Is there any procedure to add our recording.

daniestevez replied:
`satellite-recordings` is a git submodule.

PrajinIr replied:
We have also cloned satellite-recording sub module, added our  file and created a pull request.

daniestevez reviewed (changes requested):
Thanks for submitting this. Please take a look at my review comments.

daniestevez commented:
Also, just to be up-front, this Hex20 Nila satellite looks pretty much like a commercial satellite that will mis-use Amateur satellite spectrum. I didn't see any mention of Amateur radio in [Hex20's webpage](https://hex20.space/), and [this news article](https://technopark.org/technopark-based-hex20-to-fly-its-maiden-satellite-with-spacex) says "The Nila mission will qualify HEX20’s indigenously developed sub-systems and a payload for in-orbit demonstration from German company Dcubed.", while [this other article says](https://timesofindia.indiatimes.com/city/thiruvananthapuram/kerala-startup-to-make-history-satellite-nila-to-be-part-of-spacex-mission/articleshow/114969095.cms) "The Nila mission marks HEX20's entry into the hosted payload market, a key service area for the startup. According to Aravind MB, co-founder, Nila satellite will carry the payload for in-orbit demonstration from a German company, DCubed. Nila will be based on HEX20's payload-neutral satellite platform built on indigenously developed sub-systems. It will not only qualify HEX20's platform for future missions but will also help it reduce the turnaround time for building satellites in the future."

Nevertheless, I am willing to get this merged. In the [gr-satellites policy for commercial satellites](https://github.com/daniestevez/gr-satellites/blob/main/commercial_satellites.md) we have:

* "The gr-satellites project will only accept code contributions (whether in the form of SatYAML files or GNU Radio blocks) specifically intended for the reception of a commercial satellite if those contributions are endorsed or acknowledged by the satellite operator or owner.", which is the case here, since @PrajinIr is with Hex20 according to discussion in #714.

* "The gr-satellites project also acknowledges that there exist many satellites that transmit on amateur radio spectrum that are not amateur in nature, and that these satellites should not be using amateur spectrum. [...] For this reason, gr-satellites will continue trying to support any satellite that transmits on amateur radio spectrum, regardless of the nature of the mission, and accepts all contributions to this end." Which is the ultimate reason why it makes sense to merge this code from the point of view of gr-satellite's mission.

PrajinIr commented:
We have understood all the suggestions you have made regarding SatNogs and IARU. IARU has declined our request but we are in the process of adding the transmitter in the SatNogs DB. Please review this pull request and merge it, if all the corrections made on the decoder related files are satisfactory.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,232 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2025 Daniel Estevez <daniel@destevez.net>
```
I haven't written this code, so it cannot be `Copyright Daniel Estevez`.

PrajinIr replied:
We have modified the copyright snippet in the file

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,232 @@
+    'packet' / Switch(
+        lambda c: (c.hex20_header.ap_id),
+        {
+            (0x10): nila_beacon,
+            (0x60): nila_pl_pck,
```
Parenthesis are not needed. This is just a dictionary.

PrajinIr replied:
We have removed the parenthesis.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,232 @@
+    'ax25_header' / Header,
+    'hex20_header' / nila_header,
+    'packet' / Switch(
+        lambda c: (c.hex20_header.ap_id),
```
Parenthesis are not needed around the body of the `lambda`.

PrajinIr replied:
We have removed the parenthesis.

daniestevez commented:
I'm taking a look at the output of `gr_satellites Nila --wavfile satellite-recordings/Nila.wav`. Parsed telemetry looks like this:
```
-> Packet from 9k6 FSK downlink
Container: 
    ax25_header = Container: 
        addresses = ListContainer: 
            Container: 
                callsign = u'BCT' (total 3)
                ssid = Container: 
                    ch = False
                    ssid = 0
                    extension = False
            Container: 
                callsign = u'IS-1' (total 4)
                ssid = Container: 
                    ch = True
                    ssid = 0
                    extension = True
        control = 0x03
        pid = 0xF0
    hex20_header = Container: 
        sync_word_1 = 72
        sync_word_2 = 50
        sync_word_3 = 48
        version = 1
        ap_id = 16
        seq_no = 3192
        pkt_len = 143
        obc_time = 1691101520
    packet = Container: 
        obctime = 1691101520
        sat_curr_mode = (enum) NOMINAL 1
        safe_exit_thresh = 7800
        safe_enter_thresh = 7400
        battchg_volt = 8.35625
        battchg_curr = 0.0
        dep_retry_flag = 1
        pl_on_flag = 0
        oot_counter = 48
        last_cmd_opcode = 0
        cmd_accpt_count = 20
        cmd_rjct_count = 1
        bcn_flash_write_ptr = 12464
        bcn_flash_read_ptr = 0
        pl_flash_write_ptr = 4760
        pl_flash_read_ptr = 0
        ads_flash_write_ptr = 0
        ads_flash_read_ptr = 0
        uhf_flash_write_ptr = 4784
        uhf_flash_read_ptr = 0
        panel_0_curr = 0.004
        panel_1_curr = 0.004
        panel_2_curr = 0.004
        panel_0_volt = 2.416
        panel_1_volt = 2.392
        panel_2_volt = 2.432
        cdh_3v3_current = 0.108
        uhf_6v_current = 0.04
        ads_12v_current = 0.02
        cdh_3v3_volt = 3.288
        uhf_6v_volt = 6.072
        ads_12v_volt = 12.288
        pl_12v_current = 0.0
        battery_heater_current = 0.012
        main_battery_current = 0.08
        pl_12v_volt = 0.0
        battery_heater_volt = 8.36
        main_battery_volt = 8.36
        temp_solarpanel0 = -56.48924997133184
        temp_solarpanel1 = -56.48924997133184
        temp_solarpanel2 = -56.48924997133184
        temp_battery_thermistor1 = -53.69898698106138
        temp_battery_thermistor2 = -53.64132060258888
        temp_cdh = 24.135353404335472
        temp_if_card = -56.48924997133184
        temp_eps = 23.493699391500034
        actcmdrejectnum = 0
        lastrejectedactuator = 0
        dcubedplcurrentactuator = 0
        dcubedplactuatorstatus_nD3PP = 0
        dcubedplactuatorstatus_nD3RN = 0
        dcubedplactuatorstatus_uD3PP = 0
        dcubedplactuatorstatus_uD3RN = 0
        dcubedplactuatorstatus_nD3SP = 0
        dcubedplactuatorstate_nD3PP = 0
        dcubedplactuatorstate_nD3RN = 0
        dcubedplactuatorstate_uD3PP = 0
        dcubedplactuatorstate_uD3RN = 0
        dcubedplactuatorstate_nD3SP = 0
        dcubedplactuatorretrycount_nD3PP = 0
        dcubedplactuatorretrycount_nD3RN = 0
        dcubedplactuatorretrycount_uD3PP = 0
        dcubedplactuatorretrycount_uD3RN = 0
        dcubedplactuatorretrycount_nD3SP = 0
        actuateoverrideflag = 0
        frtcseconds = 165
        rtcminutes = 165
        rtchours = 165
        rtcday = (enum) (unknown) 165
        rtcdate = 165
        rtcmonth = (enum) (unknown) 85
        rtcyear = 2265
        I2Ctimeouteventcnt = 108
        squhfchannel = 65
        static_table_crc_status = (enum) TABLE_NOT_CORRUPTED 1
        dynamic_table_crc_status = (enum) TABLE_NOT_CORRUPTED 1
    crc = 48366
```

Is all this correct? Some values that look potentially wrong temperatures around -53 (Celsius I presume?) and the RTC fields.

Also:
```
$ gr_satellites Nila --wavfile /tmp/Nila.wav  | grep seq_no
        seq_no = 3192
        seq_no = 3309
        seq_no = 3387
        seq_no = 3413
        seq_no = 3465
        seq_no = 3478
```
Are the sequence numbers supposed to be consecutive? This looks like very few decoded packets. It looks like the WAV file has many more.

daniestevez commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,232 @@
+    'dcubedplactuatorretrycount_uD3RN'/Int8ul,
+    'dcubedplactuatorretrycount_nD3SP'/Int8ul,
+    'actuateoverrideflag' / Int8ul,
+    'frtcseconds' / Int8ul,
```
```suggestion
    'rtcseconds' / Int8ul,
```

PrajinIr replied:
Made the change as suggested.

daniestevez commented:
I have counted by hand 38 packets in `Nila.wav`.

daniestevez commented:
The waveform in `Nila.wav` looks quite distorted, so no wonder that decoding performance is bad. Can you provide a better recording?
![nila](https://github.com/user-attachments/assets/2c9cd50f-ecc4-43fa-b51d-77833bab25b2)

PrajinIr commented:
> I'm taking a look at the output of `gr_satellites Nila --wavfile satellite-recordings/Nila.wav`. Parsed telemetry looks like this:
> 
> ```
> -> Packet from 9k6 FSK downlink
> Container: 
>     ax25_header = Container: 
>         addresses = ListContainer: 
>             Container: 
>                 callsign = u'BCT' (total 3)
>                 ssid = Container: 
>                     ch = False
>                     ssid = 0
>                     extension = False
>             Container: 
>                 callsign = u'IS-1' (total 4)
>                 ssid = Container: 
>                     ch = True
>                     ssid = 0
>                     extension = True
>         control = 0x03
>         pid = 0xF0
>     hex20_header = Container: 
>         sync_word_1 = 72
>         sync_word_2 = 50
>         sync_word_3 = 48
>         version = 1
>         ap_id = 16
>         seq_no = 3192
>         pkt_len = 143
>         obc_time = 1691101520
>     packet = Container: 
>         obctime = 1691101520
>         sat_curr_mode = (enum) NOMINAL 1
>         safe_exit_thresh = 7800
>         safe_enter_thresh = 7400
>         battchg_volt = 8.35625
>         battchg_curr = 0.0
>         dep_retry_flag = 1
>         pl_on_flag = 0
>         oot_counter = 48
>         last_cmd_opcode = 0
>         cmd_accpt_count = 20
>         cmd_rjct_count = 1
>         bcn_flash_write_ptr = 12464
>         bcn_flash_read_ptr = 0
>         pl_flash_write_ptr = 4760
>         pl_flash_read_ptr = 0
>         ads_flash_write_ptr = 0
>         ads_flash_read_ptr = 0
>         uhf_flash_write_ptr = 4784
>         uhf_flash_read_ptr = 0
>         panel_0_curr = 0.004
>         panel_1_curr = 0.004
>         panel_2_curr = 0.004
>         panel_0_volt = 2.416
>         panel_1_volt = 2.392
>         panel_2_volt = 2.432
>         cdh_3v3_current = 0.108
>         uhf_6v_current = 0.04
>         ads_12v_current = 0.02
>         cdh_3v3_volt = 3.288
>         uhf_6v_volt = 6.072
>         ads_12v_volt = 12.288
>         pl_12v_current = 0.0
>         battery_heater_current = 0.012
>         main_battery_current = 0.08
>         pl_12v_volt = 0.0
>         battery_heater_volt = 8.36
>         main_battery_volt = 8.36
>         temp_solarpanel0 = -56.48924997133184
>         temp_solarpanel1 = -56.48924997133184
>         temp_solarpanel2 = -56.48924997133184
>         temp_battery_thermistor1 = -53.69898698106138
>         temp_battery_thermistor2 = -53.64132060258888
>         temp_cdh = 24.135353404335472
>         temp_if_card = -56.48924997133184
>         temp_eps = 23.493699391500034
>         actcmdrejectnum = 0
>         lastrejectedactuator = 0
>         dcubedplcurrentactuator = 0
>         dcubedplactuatorstatus_nD3PP = 0
>         dcubedplactuatorstatus_nD3RN = 0
>         dcubedplactuatorstatus_uD3PP = 0
>         dcubedplactuatorstatus_uD3RN = 0
>         dcubedplactuatorstatus_nD3SP = 0
>         dcubedplactuatorstate_nD3PP = 0
>         dcubedplactuatorstate_nD3RN = 0
>         dcubedplactuatorstate_uD3PP = 0
>         dcubedplactuatorstate_uD3RN = 0
>         dcubedplactuatorstate_nD3SP = 0
>         dcubedplactuatorretrycount_nD3PP = 0
>         dcubedplactuatorretrycount_nD3RN = 0
>         dcubedplactuatorretrycount_uD3PP = 0
>         dcubedplactuatorretrycount_uD3RN = 0
>         dcubedplactuatorretrycount_nD3SP = 0
>         actuateoverrideflag = 0
>         frtcseconds = 165
>         rtcminutes = 165
>         rtchours = 165
>         rtcday = (enum) (unknown) 165
>         rtcdate = 165
>         rtcmonth = (enum) (unknown) 85
>         rtcyear = 2265
>         I2Ctimeouteventcnt = 108
>         squhfchannel = 65
>         static_table_crc_status = (enum) TABLE_NOT_CORRUPTED 1
>         dynamic_table_crc_status = (enum) TABLE_NOT_CORRUPTED 1
>     crc = 48366
> ```
> 
> Is all this correct? Some values that look potentially wrong temperatures around -53 (Celsius I presume?) and the RTC fields.
> 
> Also:
> 
> ```
> $ gr_satellites Nila --wavfile /tmp/Nila.wav  | grep seq_no
>         seq_no = 3192
>         seq_no = 3309
>         seq_no = 3387
>         seq_no = 3413
>         seq_no = 3465
>         seq_no = 3478
> ```
> 
> Are the sequence numbers supposed to be consecutive? This looks like very few decoded packets. It looks like the WAV file has many more.

The decoded packet is as expected only. The RTC and temperature parameters are just default values because the RTC and solar panels are not connected during this experiment.

PrajinIr commented:
> The waveform in `Nila.wav` looks quite distorted, so no wonder that decoding performance is bad. Can you provide a better recording? ![nila](https://private-user-images.githubusercontent.com/15093841/413527711-2c9cd50f-ecc4-43fa-b51d-77833bab25b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4ODM2MTksIm5iZiI6MTczOTg4MzMxOSwicGF0aCI6Ii8xNTA5Mzg0MS80MTM1Mjc3MTEtMmM5Y2Q1MGYtZWNjNC00M2ZhLWI1MWQtNzc4MzNiYWIyNWIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDEyNTUxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5NTRjMzU3OTYwMjY5YTRmNTkwNjYyZGQ5NmMxMTFlZjM0NjBjNGE1ZTU1YmVlODQ0ZTIwMWNlMjBhMGUzODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.p6X5jX2-aEwzaqJ_LHU0ROkvy1PV-aS9RRZn_mpYk_Q)

The wav file is recorded using an engineering grade comm module only. So, the waveform may be distorted. In the actual mission we are using flight grade comm module. 
This wav file contains multiple packets of different API ids. But we are interested in just beacon packet and payload packet decoding.

daniestevez commented:
> The wav file is recorded using an engineering grade comm module only. So, the waveform may be distorted. In the actual mission we are using flight grade comm module.

That doesn't make much sense, but in any case, let's not commit a distorted recording to the `satellite-recordings` submodule. We can proceed without a recording if you cannot provide a good sample recording.

daniestevez commented:
> This wav file contains multiple packets of different API ids. But we are interested in just beacon packet and payload packet decoding.

That raises a good point. If there are other APIDs besides `0x10` and `0x60`, then a `default=GreedyBytes` ([example](https://github.com/daniestevez/gr-satellites/blob/1410dbd3a975cee84000348313a8026f6686debd/python/telemetry/aepex_70cm.py#L42)) should be added so that packets from other APIDs do not cause a telemetry parser failure.

PrajinIr commented on python/telemetry/nila.py:
```diff
@@ -0,0 +1,232 @@
+        {
+            0x10: nila_beacon,
+            0x60: nila_pl_pck,
+        }
```
We have added the GreedyBytes.

PrajinIr commented:
> > The wav file is recorded using an engineering grade comm module only. So, the waveform may be distorted. In the actual mission we are using flight grade comm module.
> 
> That doesn't make much sense, but in any case, let's not commit a distorted recording to the `satellite-recordings` submodule. We can proceed without a recording if you cannot provide a good sample recording.

We have removed the recording from the satellite-recordings folder.

daniestevez commented:
This looks good now. Can you squash all the commits before I merge?

PrajinIr commented:
> This looks good now. Can you squash all the commits before I merge?

I have squashed the commits and pushed.
Please merge.

daniestevez commented:
> I have squashed the commits and pushed.

This is not squashed. There are 5 commits in the PR branch. There should be only one commit.

PrajinIr commented:
We have started the work from a previous commit, not from the latest main branch.
So once squash is created, it from head behind only.

Shall I create a new branch from the latest main and cherry pick all the changes to the new branch and make a single commit.
If its ok,I will generate the pull request from there.

daniestevez commented:
Please squash all the commits into a single commit and `push --force` to this PR branch. Do not create a new pull request.

daniestevez commented on test.sh:
```diff
@@ -2,6 +2,7 @@
 
 export GR_SATELLITES_SUBMIT_TLM=0
 
+
```
Spurious spacing change in this file.

daniestevez commented:
The git SHA of the `satellite-recordings` submodule should not be changed by this PR, since no files have actually been added.

PrajinIr commented:
> The git SHA of the `satellite-recordings` submodule should not be changed by this PR, since no files have actually been added.

I plan to use this command in git repo "satelitte-recordings "
Will this be a solution

PrajinIr commented:
> > The git SHA of the `satellite-recordings` submodule should not be changed by this PR, since no files have actually been added.
> 
> I plan to use this command in git repo "satelitte-recordings " Will this be a solution

git reset --hard b77f429c39890e64c4d9924d08ec9f675018c7af

daniestevez commented:
> Will this be a solution

Probably not. Nothing that you do inside `satellite-recordings` would change how a commit in `gr-satellites` looks like.

PrajinIr commented:
> > Will this be a solution
> 
> Probably not. Nothing that you do inside `satellite-recordings` would change how a commit in `gr-satellites` looks like.

Can you please suggest a solution. Sorry for the inconvenience caused

PrajinIr commented:
> > > The git SHA of the `satellite-recordings` submodule should not be changed by this PR, since no files have actually been added.
> > 
> > 
> > I plan to use this command in git repo "satelitte-recordings " Will this be a solution
> 
> git reset --hard b77f429c39890e64c4d9924d08ec9f675018c7af

I have tested this and worked in our repo.
I have followed the below sequence.
1. From the latest commit in a branch, I have added a file.
2. Then I commit the change and pushed to repo.
3. I have removed the file.
4. Then I commit the change and pushed to repo.
5. I have reset the git repo using the command, "git reset --hard xxxxxxxxxxxxxxxxxxxx"
6. Then I again pushed.
7. Remote git HEAD fallback to the given commit hash as expected.

This is expected to be reflected in repo "satellite-recordings".

Shall we proceed with the last change in the test.sh and squash sequence in "gr-satellites" repo after reset the satellite-recordings?

daniestevez commented:
I have merged this manually in 672df323242640ee4537580d523a4cd2c454e143. Closing this PR.
